### ISSUE TITLE
Option on $location to allow hash/path change w/o reloading the route

### ISSUE THREAD

**Issue by cgross** [angular/angular.js#1699]
We have a paradigm in our app that user's create new things on the same page that they view things.  So our route is like /thing/:id.  When creating a new thing they go to /thing/new.  Once the thing has been successfully saved we want to change the route to /thing/1234 (or whatever its new id is).  Now the partial doesnt need to be reloaded because the data is all the same.  We just want the path to be updated so a user can now bookmark the correct link, etc.

Having an option on $location (not on the route definition) to enable/disable route loading would work but I'm sure there are other ways to implement the feature.

**Comment by pankajkumar** [comment]
+1

**Comment by macinnir** [comment]
+1

**Comment by marcocom** [comment]
+1 what about a completely seperate method?  leaving path() as is, and just adding a $location.hash method that simply changes the browser's url without controller reload.  this allows us to control back-navigation of the browser and is a valid use-case scenario in rich applications.

**Comment by pencilcheck** [comment]
+1

Currently using the snippet here in this page: http://joelsaupe.com/programming/angularjs-change-path-without-reloading/

**Comment by benbeadle** [comment]
+1

**Comment by zuzusik** [comment]
+1

**Comment by briansanchez** [comment]
+1

**Comment by vegetableman** [comment]
:+1:

**Comment by van-nguyen** [comment]
A workaround I found was to add a child state with an absolute url.

    $stateProvider.state('state.child', {
        url: '^/some-other-url'
    });

Then simply do $state.go('.child') and that should change the url to the one you want without reloading and keep everyone happy.

**Comment by AreYouSiries** [comment]
+1

**Comment by manikantag** [comment]
+1

**Comment by tsl0922** [comment]
+1

**Comment by MarkoCen** [comment]
+1

**Comment by swkruger** [comment]
+1

**Comment by MartelBenjamin** [comment]
+1

**Comment by codler** [comment]
+1

**Comment by petebacondarwin** [comment]
@btford how is this dealt with (if at all) in the `componentRouter`?

**Comment by iki** [comment]
Thanks @balsarori for [a great solution](https://github.com/angular/angular.js/issues/1699#issuecomment-45048054). Works great!

Elaborated on top of it to allow updating $route path parameters without reloading the controller:
```js
'use strict';

angular.module('params', [
  'ng',
  'ngRoute',
])
.factory('params', function($route, $location) {
  var service = {};

  /**
   * @ngdoc method
   * @name params#replace
   *
   * @description
   * Replace route params, including path params, without reloading controller.
   *
   * Causes `$route` service to update the current URL, replacing
   * current route parameters with those specified in `params`.
   *
   * Provided property names that match the route's path segment
   * definitions will be interpolated into the location's path, while
   * remaining properties will be treated as query params.
   *
   * If `opts.reload` is truthy, the current controller is reloaded.
   *
   * If `opts.history` is truthy, the browser history is updated.
   * If `opts.history` is undefined, the browser history is only updated
   * when changing value of any existing `$route` parameter,
   * ie. not when adding a previously undefined `$route` parameter.
   *
   * @param {!Object<string, string>} params mapping of URL parameter names to values
   * @param {!Object<string, boolean>} opts `history` and `reload` options.
   *
   * @returns {Object} self
   */
  service.replace = function(params, opts) {
    // See https://github.com/angular/angular.js/issues/1699#issuecomment-45048054
    // and https://github.com/angular/angular.js/tree/v1.3.x/src/ngRoute/route.js#L484
    // and https://github.com/angular/angular.js/tree/v1.3.x/src/ngRoute/route.js#L539
    var key;

    if (!opts || !opts.reload) {
      var path = $route.current.$$route.originalPath;

      for (key in params) {
        // console.log('params.replace: ' + (path.search(':\\b' + key + '\\b') ? 'path.' : 'search.') +
        //    key + ' = ' + (key in $route.current.params ? $route.current.params[key] + ' -> ' : '') + params[key]);

        if (path.search(':\\b' + key + '\\b') !== -1) {
          $route.current.pathParams[key] = '' + params[key];
        }
      }
    }

    if (!opts || !('history' in opts)) {
        var current = $route.current.params;

        for (key in params) {
          if (key in current && params[key] !== current[key]) {
            console.log('params.replace: update history: ' + key + ' = ' + current[key] + ' -> ' + params[key]);
            (opts || (opts={})).history = true;
            break;
          }
        }
    }
    if (!opts || !opts.history) {
        $location.replace();
    }

    $route.updateParams(params);

    return service;
  };

  return service;
});
```

**Comment by marcalj** [comment]
What's the current status of this issue? It will be solved in the new router? On-time for 1.4 release? Soonish?

I'm having the same problem but with query string parameters. I don't want a reload when changing them. But I can't disable ```reloadOnSearch``` because I need to we aware if the user changes the query string manually! :)

Thanks!

**Comment by irthos** [comment]
@marcalj fwiw, we hacked this solution with a service that kept the url in sync with the address bar so that we could disable `reloadOnSearch` and grab query parameters. also, users can navigate the url directly via bookmark or link.

**Comment by marcalj** [comment]
With ng-router or ui-router? Can you provide me a link with the code or something?

Maybe this could be solved using new router, I can wait for it since my project it's not finished yet.

Thanks!

**Comment by irthos** [comment]
@marcalj our example is a bit convoluted and possibly outdated but look at https://github.com/irthos/mngr/blob/master/app.js and check line 10. it uses our api.loadState method which identifies params in the url to construct the proper state.

**Comment by jicasada** [comment]
NgSilent was the solution for me.
Look at https://github.com/garakh/ngSilent

**Comment by Siraris** [comment]
I used both @sergey-buturlakin  and @kuchumovn solution, and I continually run into the same issue.

On the page in question (pagex) I use location.skipReload and it works great - I can update the path over and over without a route change.  If I then use an actual $location.path to change the route away from pagex to a new controller, it works, but if I hit the back button to return to pagex, and then try to navigate to another page, I can no longer change my route anymore without refreshing the browser.  Is there a way around this?

**Comment by harmjanluth** [comment]
+1

**Comment by sai571** [comment]
+1. I like double colon approach that Matias suggested

**Comment by kressh** [comment]
+1

**Comment by garmoshka-mo** [comment]
It is common use case for single page app to not reload page on path change,
very strange why it is not yet embedded to angular.
made plugin based on @EvanWinstanley solution:
https://github.com/anglibs/angular-location-update

**Comment by petebacondarwin** [comment]
Let's revisit this in 1.5. There is a related PR at #5860

**Comment by marcalj** [comment]
I need the same function for query string update without triggering route controller! :)

Something like: ```$location.searchSilent('sf', SearchFilter.getForRouteParams());```

**Comment by jlmayoral** [comment]
$location.path(href);  
  scope.$apply();
is the solution!!
we need inform to angular the change with $apply()

**Comment by ajwhite** [comment]
This thread has not been open for 2 years because angular hadn't been notified -- we want to change a part of the route without invoking everything.

**Comment by iki** [comment]
@petebacondarwin it would be good to have not only location path update without reload, but also path params update without reload ... like implemented above https://github.com/angular/angular.js/issues/1699#issuecomment-96126887

**Comment by peterwidmer** [comment]
The solution of @jicasada worked perfectly fine for me! Thx!

**Comment by shavo007** [comment]
@jicasada does this work with ui-router? I have raised a bug with ui-router where reloadOnSearch:false does not work if parent and child state share the same URL. If I transition to child state with updated request param it will call the parent controller twice.

**Comment by Luntegg** [comment]
+1

**Comment by pbouda** [comment]
+1

**Comment by vinogradov** [comment]
+1

**Comment by Krisa** [comment]
I confirm that ngSilent is doing perfectly the job. Even so an official solution would be highly appreciated.

**Comment by freeman14** [comment]
The solution is essentially to add an extra parameter to $location.path().
```
app.run(['$route', '$rootScope', '$location', function ($route, $rootScope, $location) {
    var original = $location.path;
    $location.path = function (path, reload) {
        if (reload === false) {
            var lastRoute = $route.current;
            var un = $rootScope.$on('$locationChangeSuccess', function () {
                $route.current = lastRoute;
                un();
            });
        }
        return original.apply($location, [path]);
    };
}]);
```
And then just use like this
```
$location.path('/path/that/you/need', false);
```

**Comment by JohnGalt1717** [comment]
I should just be able to do $location.replace(<somepath>) just like the window.location.replace function. This is just silly because every Create in a CRUD operation will always have this issue.

**Comment by stitisami** [comment]
(y) +1

**Comment by espeandr** [comment]
+1

**Comment by yalishizhude** [comment]
I advise you use "search param" like this:
` /thing?id=1234`
Do not config the param in the router provider,it won't reload view and controller.
If you want to get **id**
`$location.$$search().id //1234`
@cgross

**Comment by dolymood** [comment]
Demo http://codepen.io/dolymood/details/jrEQBx/
Use `decorator` to decorator `ngViewDirective`

**Comment by mattantonelli** [comment]
+1

**Comment by shrijanshukla89** [comment]
+1

**Comment by Meokme** [comment]
+1

**Comment by nakul12345** [comment]
testing teting testing

**Comment by nakul12345** [comment]
testing testing

**Comment by nakul12345** [comment]
testing

**Comment by nakul12345** [comment]
testing

**Comment by nakul12345** [comment]
testing

**Comment by nakul12345** [comment]
test

**Comment by nakul12345** [comment]
tesdfsdfs

**Comment by nakul12345** [comment]
yrdfdfgdfgdfg

**Comment by nakul12345** [comment]
sdfsfsdf

**Comment by bytexro** [comment]
+1

**Comment by b-mcrae** [comment]
+1

**Comment by nitinsurana** [comment]
+1

**Comment by risingfish** [comment]
+!

**Comment by petebacondarwin** [comment]
Enough of the +1s already :-)
We get the message. We will look at this before Xmas!

**Comment by gkalpak** [comment]
Closed by #15002.
